ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -889,28 +901,97 @@ static bool TryGetVirtualizationRoot(
         *kauthResult = KAUTH_RESULT_DEFER;
         return false;
     }
-    else if (!VirtualizationRoot_IsOnline(*root))
+    
+    return true;
+}
+
+
+static bool CurrentProcessIsOwnedOrWasSpawnedByRegularUser()
+{
+    bool nonServiceUser = false;
+    
+    proc_t process = proc_self();
+    
+    while (true)
+    {
+        kauth_cred_t credential = kauth_cred_proc_ref(process);
+        uid_t processUID = kauth_cred_getuid(credential);
+        kauth_cred_unref(&credential);
+        
+        if (processUID >= 500)
+        {
+            nonServiceUser = true;
+            break;
+        }
+        
+        pid_t parentPID = proc_ppid(process);
+        if (parentPID <= 1)
+        {
+            break;
+        }
+        
+        proc_t parentProcess = proc_find(parentPID);
+        proc_rele(process);
+        process = parentProcess;
+        if (parentProcess == nullptr)
+        {
+            KextLog_Error("CurrentProcessIsOwnedOrWasSpawnedByRegularUser: Failed to locate ancestor process %d for current process %d\n", parentPID, proc_selfpid());
```

Comment:
Is it possible that the oldest ancestor process has a PID between 2 and 499?

Reply:
The ultimate parent is, as far as I'm aware, always `launchd` (PID 1), which runs as root (UID 0). Does that answer the question? If not, please elaborate!

Reply:
> The ultimate parent is, as far as I'm aware, always launchd (PID 1), which runs as root (UID 0). Does that answer the question? If not, please elaborate!

Yep that answers my question thanks!  I was trying to understand the reason for the gap between 1 and 500 in the loop, but I see now that we just walk all the way up the process tree and bail if/when we find a user-spawned process (>= 500).